src/style/style_layer/line_style_layer.js

```diff
@@ -59,14 +61,19 @@ class LineStyleLayer extends StyleLayer {
 
     _handleSpecialPaintPropertyUpdate(name: string) {
         if (name === 'line-gradient') {
-            this._updateGradient();
+            const expression: ZoomConstantExpression<'source'> = ((this._transitionablePaint._values['line-gradient'].value.expression): any);
+            this.stepInterpolant = expression._styleExpression.expression instanceof Step;
```

Comment:
There are cases where the top-level expression is not directly a `step` expression, even though this expression does not support dds.

For example:
```
['let', 
'my_var', ['+', 1, 1],
['step', .....
```
While, this is not required, it is allowed by the expression system.

Reply:
That's a good point, I will test this part to better account for detecting the expression.

Reply:
I've tried a few different combinations of the above to write a test for it but couldn't the expression to accept it, but I could be missing something. With what I tried, `line-gradient` seemed to be fairly restrictive in what it can ingest as input ([`ColorRampProperty`](https://github.com/mapbox/mapbox-gl-js/blob/5edbaf1a8186dfa0aad8c955ee958c09492ec855/src/style/style_layer/line_style_layer_properties.js#L63)): `layers.line.paint.line-gradient: color expected, array found`. We might be safe here.